text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,206 @@
+
+```bash
+.
+├── buildpack.toml
+├── cmd
+│   └── main.go
+├── go.mod
+├── go.sum
+└── pkg
+    ├── build.go
+    └── detect.go
+```
+
+This approach is not opinionated on topics such as which testing framework to use.  Only components under github.com/buildpacks/* are used in the generated project.  `pack buildpack create --template URL` allows for more opinionated project scaffolding.  As such, we support new buildpack authors with onboarding and support experienced buildpack authors to use the scaffolding of their choice.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The design is modeled on [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/) with reference to [springerle](https://github.com/carlmjohnson/springerle) -- a similar implementation in golang -- and [create-go-app](https://github.com/create-go-app/).  We do we want to `os.Exec` a python subprocess to run cookiecutter as this would require availability of a python runtime environment.  Instead we propose to borrow heavily from create-go-app, and generate the default shell and libcnb skeletons from a cloned git repoistory.
+
+A full invocation to generate `bash` scaffolding prompts for the values [currently documented as flags](https://buildpacks.io/docs/buildpack-author-guide/create-buildpack/building-blocks-cnb/#using-the-pack-cli).
+
+```bash
+pack buildpack new
+Use the arrow keys to navigate: ↓ ↑ → ←
+? Choose a project template:
+  ▸ bash
+    go
+
+```
+
+A full session includes the terminal prompts that the project scaffolding tool asks of the end user:
+
+```bash
+pack buildpack new
+Use the arrow keys to navigate: ↓ ↑ → ←
+? Choose a project template:
+  ▸ bash
+    go
+✔ Enter a directory in which to scaffold the project: bash_buildpack
+Use the arrow keys to navigate: ↓ ↑ → ←
+? Choose the buildpack API version (use the default if you are unsure):
+  ▸ 0.7
+    0.8
+✔ Enter an identifier for the buildpack: example/bash
+✔ Enter the initial buildpack version: 0.0.1
+✔ Enter a stack this buildpack will support by default: io.buildpacks.samples.stacks.bionic
+
+Created project in bash_buildpack
+```
+
+Templates are provided as a file tree.  A root-level `prompts.toml` file contains prompts for the end-user.  For example, a partial file tree containing user prompts and Go skeleton code is:
+
+```bash
+.
+├── prompts.toml
+└── {{.ProjectDirectory}}
+    ├── buildpack.toml
+    ├── cmd
+    │   └── main.go
+    ├── go.mod
+    └── pkg
+        ├── build.go
+        └── detect.go
+```
+
+Where `prompts.toml` is of the form:
+
+```toml
+[[prompt]]
+name="ProjectDirectory"
+prompt="Enter a directory in which to scaffold the project"
+default="bash_buildpack"
+required=true
```

Comment:
Would it be possible to have certain prompts that are always displayed? Like project directory, buildpack API, etc.. 

It seems repetitive to have to set them in every template, when pretty much every template is going to need that info.

Reply:
The current `pack buildpack new` has two flags `--api` and `--path`.  We could maintain support for these flags in `pack buildpack create`?

Reply:
+1 for that. It'll just reduce some of the repetition for template authors.

Reply:
I think we decided that we would not want that. There can definitely be templates that don't need to prompt for the api version and just support one.

Reply:
True. I see ways around that though.

1. Pack could look at the template and see if it's using that variable. If it's not, then skip the question.

2. It wouldn't be hard for pack to augment the question list though. If pack takes the list of input required from the template, adds to that any default questions, and filters out the ones the template has overridden you could without too much trouble allow a standard list of questions but let templates override them.

    The template that only wants to support one API version could do so and set the allowed list to one item. Pack could then ask the question with only a single answer or we could perhaps optimize and skip that question knowing the answer. Or we could get fancy and add a "hidden" property to allow the template to control if it wants that to be shown or not.

It's a little extra effort but I think it could be worth it.

Reasons:

1. Templates will get out of date. That'll never happen to the official templates ;) but community templates which I hope crop up will need to be maintained to have things like the API updated when new API versions come out. That could be annoying for users, especially if the API version is formatted as a select list with no "other" option.

2. If you have many templates to update that would be a pain, especially if they are across multiple repos (I think it was asked but haven't seen an answer regarding if you can have multiple templates in a single git repo, that would lessen this problem).

Reply:
I think the core problem here is coupling between `pack` binary and "official" scaffolding.  Previously the scaffolding was distributed with  `pack`.  We're somewhat decoupling `pack` and official scaffolding.  However, we see situations where an older `pack` that supports only (for example) API `[0.7, 0.8]` might call out to newer official scaffolding that _can_ generate projects spec'ed to `1.0`.

One way of solving this is for `pack` to provide to the scaffolding, the set of API versions that `pack` supports.  It would then be the responsibility of the scaffolding to generate compliant projects.  This is what the current proposal suggests.  `pack` can provide a list of `BuildpackAPI` versions to the scaffolding.  `pack` can either choose to use one of these versions (an override) or leave it to the end-user to be prompted for one of these versions.

I wonder if there are other ways to solve this?  Should we more tightly couple `pack` and the official scaffolding?  So `pack` would point at a specific release of the official scaffolding?  This way an older version of `pack` will never try to use a newer scaffolding.  Furthermore, rather than coupling `pack` releases to scaffolding releases, we could release scaffolding by API version?   Thus `pack` supporting  API `[0.7, 0.8]` would scaffold using the latest release of scaffolding for API version `0.7` or API version  `0.8`.  This allows bug-fixes in the `0.8` release to be available to all `pack` releases that support API version `0.8`.

The key issue appears to be `pack` telling the scaffolding the API versions that it supports.  We can discuss this more widely in a BAT meeting.

Reply:
We've taken the view that this is not an issue that needs to be addressed right now.